text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+
+At a high level, we define the expected customer experience when upgrading either of the two CDK components:
+
+- CLI upgrades are compatible. New CLI versions should work with older framework versions, and all existing functionality is preserved.
+- Framework upgrades are incompatible. We will require the user to upgrade the CLI as well.
+
+This will eventually lead the user to always have a setup where `CLI >= Framework`.
+
+# Motivation
+
+The motivation behind this RFC is three-fold:
+
+## Confusing Upgrade Process
+
+Currently, the upgrade process for our users can be a bit complicated/confusing.
+
+We generally don't require users to upgrade their CLI when they upgrade the framework.
+
+However, sometimes, we introduce features in the framework that rely on new CLI capabilities.
+When this happens, we bump the cx-protocol version, forcing (all) users to upgrade their CLI as well, even those who don't use the new feature.
+
+Things get much worse, if we forget to bump the cx-protocol version (which we do sometimes).
+In this case, the user will not receive the upgrade instruction, and will rightfully assume the new feature simply works.
+While in reality, this can either cause unexpected errors or non-working features.
+
+We want to come up with a consistent model that users can rely on.
+
+## Simplify CX Protocol Versioning
+
+Recently, we have had a few customer impacting issues that relate to our cx-protocol.
+
+- [aws/aws-cdk: A newer version of the CDK CLI (>= 1.21.0) is necessary to interact with this app](https://github.com/aws/aws-cdk/issues/5986)
+- [aws/aws-cdk: CDK CLI can only be used with apps created by CDK >= 1.10.0](https://github.com/aws/aws-cdk/issues/4294)
+
+A major contributing factor to those issues, is that the ergonomics of introducing breaking changes (cx-protocol version bump), are not very good.
+Something about the process feels convoluted and we seem to be getting it wrong almost every time we attempt it.
+
+We should consider a re-think of the entire process, as well as the assumptions that lead to it.
+
+## Validate Compatibility
+
+We want to make sure we have sufficient testing in place that validate we don't introduce any accidental CLI compatibility breakage.
+
+**Currently, the only way for us to detect these types of problems is by noticing them in the review process.**
+
+# Design Summary
+
+Assuming we adopt the new compatibility model, all we have to do is come up with a validation mechanism for the following breakage scenarios:
+
+## Breaking changes in CX Protocol
+
+Probably the most prominent scenario for causing breakage is pushing an incompatible change to the cx-protocol itself.
+This can, for example, be the removal of a property.
+
+> See concrete [example](#rename-target-property-in-containerImageAssetMetadataEntry).
+
+The proposed solution for this case is to run API compatibility checks using `jsii-diff`.
```

Comment:
API compatibility will not cover details of the protocol that are *not* API (file system contracts, ...)

Reply:
You mean like asset location for instance?

You're right, I think its pretty analogous to our regular framework packages. In those, we have compatibility checks to ensure API didn't break, and we have integration tests to ensure the API output didn't change (expected CloudFormation template).

So maybe its a matter of adding integration tests to the `@aws-cdk/core` package, where the expected output is a cloud assembly directory.

Reply:
Im going to add this idea to the future possibilities section. Also, I will add a section about implementing CLI integration tests that run on previous framework versions, this will make sure cloud assemblies created by older frameworks, **works** with the new CLI, and not just API compatible.

Reply:
@RomainMuller 

https://github.com/aws/aws-cdk-rfcs/blob/epolon/compatibility-strategy/text/00110-cli-framework-compatibility-strategy.md#breaking-changes-in-cli

https://github.com/aws/aws-cdk-rfcs/blob/epolon/compatibility-strategy/text/00110-cli-framework-compatibility-strategy.md#cx-protocol-integration-tests

Reply:
My point here was mostly that `jsii-diff` can give you a "last resort" validation. It's a good safety net, but it CANNOT be deemed sufficient by itself: if the API (signatures) do not change, but the behavior of the implementation does, it's still broken but `jsii-diff` will give it an A++ pass mark.

Reply:
Absolutely correct. Thats why we have two additional mechanism that run integration tests, which is what i pointed you at. As well as another mechanism we implement in the future.